lib/livebook/hub/fly.ex

```diff
@@ -0,0 +1,124 @@
+defmodule Livebook.Hub.Fly do
+  @moduledoc false
+
+  defstruct [:id, :name, :deployed, :hostname, :organization, :image_details, :state, :token]
+
+  alias Livebook.Utils.HTTP
+
+  def fetch_applications(access_token) do
+    headers = [{"Authorization", "Bearer #{access_token}"}]
+    body = {"application/json", Jason.encode!(%{query: query()})}
+
+    with {:ok, body} <- request(headers, body) do
+      {:ok, decode_nodes(body, access_token)}
+    end
+  end
+
+  defp decode_nodes(%{"apps" => %{"nodes" => nodes}}, token) do
+    for node <- nodes do
+      org = node["organization"]
+      image = node["latestImageDetails"]
+
+      %__MODULE__{
+        id: node["id"],
+        name: node["name"],
+        deployed: node["deployed"],
+        hostname: node["hostname"],
+        organization: %{
+          id: org["id"],
+          name: org["name"],
+          slug: org["slug"],
+          type: org["type"]
+        },
+        image_details: %{
+          repository: image["repository"],
+          registry: image["registry"],
+          tag: image["tag"]
+        },
+        state: node["state"],
+        token: token
+      }
+    end
+  end
+
+  defp graphql_endpoint do
+    Application.get_env(:livebook, :fly_io_graphql_endpoint, "https://api.fly.io/graphql")
+  end
+
+  defp request(method \\ :post, url \\ graphql_endpoint(), headers, body) do
+    HTTP.request(method, url, headers: headers, body: body)
+    |> decode_body()
+    |> handle_response()
+  end
+
+  defp decode_body({:ok, 200, _, body}), do: {:ok, Jason.decode!(body)}
+  defp decode_body({:ok, _, _, body}), do: {:error, Jason.decode!(body)}
+  defp decode_body({:error, _} = error), do: error
+
+  defp handle_response({:ok, %{"errors" => [error]}}) do
+    case error do
+      %{"extensions" => %{"code" => code}} ->
+        reason =
+          code
+          |> Macro.underscore()
+          |> String.to_atom()
```

Comment:
It's best to avoid dynamic atom creation, in this case I think we should be fine normalizing the message into a string (which should also be easier to handle by the caller). Something like `"request failed with code: #{code}"`.